client/src/controllers/DrilldownController.test.js

```diff
@@ -0,0 +1,201 @@
+
+    application = new Application();
+    application.handleError = (error, message) => {
+      errors.push({ error, message });
+    };
+    application.register(identifier, DrilldownController);
+    application.register('w-action', ActionController);
+    application.register('w-dropdown', DropdownController);
+    application.start();
+
+    await jest.runAllTimersAsync();
+
+    return [
+      ...document.querySelectorAll(`[data-controller~="${identifier}"]`),
+    ].map((element) =>
+      application.getControllerForElementAndIdentifier(element, identifier),
+    );
+  };
+
+  beforeEach(async () => {
+    await setup();
+  });
+
+  afterEach(() => {
+    application?.stop && application.stop();
+    errors = [];
+    eventNames.forEach((name) => {
+      events[name] = [];
+    });
+  });
+
+  describe('DrilldownController open/close actions', () => {
+    it('opens the submenu and sets active submenu value when toggle is clicked', async () => {
+      const toggleButton = document.querySelector(
+        '[data-w-drilldown-target="toggle"]',
+      );
+      const submenu = document.querySelector('#drilldown-field-0');
+
+      toggleButton.click();
+      await jest.runAllTimersAsync();
+
+      expect(toggleButton.getAttribute('aria-expanded')).toBe('true');
+      expect(submenu.hidden).toBeFalsy();
+    });
+
+    it('closes the submenu and clears active submenu value when close action is triggered', () => {
+      const toggleButton = document.querySelector(
+        '[data-w-drilldown-target="toggle"]',
+      );
+      toggleButton.click();
+
+      const closeButton = document.querySelector(
+        '[data-action="click->w-drilldown#close"]',
+      );
+      closeButton.click();
+
+      const submenu = document.querySelector('#drilldown-field-0');
+      expect(toggleButton.getAttribute('aria-expanded')).toBe('false');
+      expect(submenu.hidden).toBeTruthy();
+    });
+  });
+
+  describe('DrilldownController delayedClose method', () => {
+    it('should call close after 200ms delay', () => {
+      document.body.innerHTML = `
+      <div id="drilldown-field-0">
+        <button data-action="w-dropdown:hide->w-drilldown#delayedClose">Close</button>
+      </div>`;
+
+      const submenu = document.querySelector('#drilldown-field-0');
+      const delayCloseButton = document.querySelector(
+        '[data-action="w-dropdown:hide->w-drilldown#delayedClose"]',
+      );
+      submenu.setAttribute('data-w-drilldown-active', 'true');
+
+      expect(delayCloseButton).not.toBeNull();
+      expect(submenu.getAttribute('data-w-drilldown-active')).toBe('true');
+
+      delayCloseButton.click();
+      expect(submenu.getAttribute('data-w-drilldown-active')).toBe('true');
+
+      jest.advanceTimersByTime(200);
+      expect(submenu.getAttribute('data-w-drilldown-active')).toBe('');
+    });
+  });
+
+  describe('DrilldownController - Action Outlet Connection/Disconnection', () => {
+    let actionElement;
+    let controller;
```

Comment:
I am sure you are on it, but this test block will need to be updated to avoid accessing the controller directly.

For the disconnection, you could just use the `events` and make sure no events are fired after removal?

Reply:
Yes, After connecting, Ensure the event listener is added correctly. And after disconnection, Ensure that nothing happens when you trigger the event (no additional actions are triggered after removing the element).